server/sonar-server/src/main/java/org/sonar/server/platform/Platform.java

```diff
@@ -276,10 +309,133 @@ public Object getComponent(Object key) {
   }
 
   public enum Status {
-    BOOTING, SAFEMODE, UP
+    BOOTING, SAFEMODE, STARTING, UP
   }
 
   public enum Startup {
     NO_STARTUP_TASKS, ALL
   }
+
+  public interface AutoStarter {
+    /**
+     * Let the autostarted execute the provided code.
+     */
+    void execute(Runnable startCode);
+
+    /**
+     * This method is called by executed start code (see {@link #execute(Runnable)} has finished with a failure.
+     */
+    void failure(Throwable t);
+
+    /**
+     * This method is called by executed start code (see {@link #execute(Runnable)} has finished successfully.
+     */
+    void success();
+
+    /**
+     * Indicates whether the AutoStarter is running.
+     */
+    boolean isRunning();
+
+    /**
+     * Requests the startcode (ie. the argument of {@link #execute(Runnable)}) aborts its processing (if it supports it).
+     */
+    void abort();
+
+    /**
+     * Indicates whether {@link #abort()} was invoked.
+     * <p>
+     * This method can be used by the start code to check whether it should proceed running or stop.
+     * </p>
+     */
+    boolean isAborting();
+
+    /**
+     * Called when abortion is complete.
+     * <p>
+     * Start code support abortion should call this method once is done processing and if it stopped on abortion.
+     * </p>
+     */
+    void aborted();
+  }
+
+  private abstract static class AutoStarterRunnable implements Runnable {
+    private final AutoStarter autoStarter;
+
+    AutoStarterRunnable(AutoStarter autoStarter) {
+      this.autoStarter = autoStarter;
+    }
+
+    @Override
+    public void run() {
+      try {
+        doRun();
+      } catch (Throwable t) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1181)